File: 960-DeleteColumnstoMakeSortedIII/960-DeleteColumnstoMakeSortedIII.py
# Last updated: 22/12/2025, 3:45:59 pm
1class Solution:
2    def minDeletionSize(self, st: List[str]) -> int:
3        n = len(st)
4        m = len(st[0])
5
6        @cache
7        def rec(index, prev_index, prev_prev_index):
8            if index == m: return 0
9            
10            ans = rec(index + 1, prev_index, prev_prev_index) + 1
11            for i in range(n):
12                if st[i][index] < st[i][prev_index]:
13                    if prev_prev_index != -1:
14                        for i2 in range(n):
15                            if st[i2][index] < st[i2][prev_prev_index]:
16                                return ans
17                    return min(ans, rec(index + 1, index, prev_prev_index) + 1)
18            
19            return min(ans, rec(index + 1, index, prev_index))
20        
21        ans = rec(1, 0, -1)
22        rec.cache_clear()
23        return ans

File: 960-DeleteColumnstoMakeSortedIII/960-DeleteColumnstoMakeSortedIII.cpp
// Last updated: 22/12/2025, 3:59:28 pm
1int dp[101][101][102];
2
3class Solution {
4public:
5    bool isCurrentIndexLargerThanPrev(int index, int prev_index, vector<string>& strs, int& n) {
6        for (int i=0; i<n; i++) {
7            if (strs[i][index] < strs[i][prev_index]) return false;
8        }
9        return true;
10    }
11
12    int rec(int index, int prev_index, int prev_prev_index, vector<string>& strs, int& n, int& m) {
13        if (dp[index][prev_index][prev_prev_index] != -1) return dp[index][prev_index][prev_prev_index];
14        if (index == m) return 0;
15
16        int ans = rec(index + 1, prev_index, prev_prev_index, strs, n, m) + 1;
17        if (isCurrentIndexLargerThanPrev(index, prev_index, strs, n)) {
18            ans = min(ans, rec(index + 1, index, prev_index, strs, n, m));
19        } else if (prev_prev_index == 101 || isCurrentIndexLargerThanPrev(index, prev_prev_index, strs, n)) {
20            ans = min(ans, rec(index + 1, index, prev_prev_index, strs, n, m) + 1);
21        }
22        dp[index][prev_index][prev_prev_index] = ans;
23        return ans;
24    }
25
26    int minDeletionSize(vector<string>& strs) {
27        int n = strs.size();
28        int m = strs[0].length();
29        memset(dp, -1, sizeof(dp));
30        return rec(1, 0, 101, strs, n, m);
31    }
32};